ColorPicker: Remove horizontal scrollbar when using HSL or RGB color input types

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Raised in the review for https://github.com/WordPress/gutenberg/pull/41222#pullrequestreview-985829772:

In the ColorPicker component, ensure that the range controls in their 100% position don't cause the component to overflow and create a horizontal scrollbar when used in a popover.

Also, ensure that the bottom margin / padding of the input fields is consistent between each color input type.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It appears that the `RangeControl` component's slider handle (intentionally) overflows the width of the control so that the handle is more visible — the drawback is that this can cause this particular Popover to overflow. To fix this, we can give the component a little more breathing room. It seemed safer to fix it here, than to update the `RangeControl` directly, where it could adversely affect other instances of that component.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a little more wrapping to the color input control components so that we can set different padding for the header area versus the input fields.
* Set a larger amount of padding and adjust the margin for the RangeControl sliders for the HSL and RGB inputs so that the 100% position doesn't overflow the container.
* Update the bottom padding to ensure that in its full height size, there's no vertical scrollbar, and that the bottom padding is consistent between Hex, RGB, and HSL inputs.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

To more clearly see when there's a horizontal scrollbar, if you're running macOS, go to System Preferences > General > Show scroll bars, and set it to Always.

1. Before this PR, open up the post editor and add a paragraph.
2. Go to set a background color and set a custom color.
3. Switch between the different kinds of color inputs, and notice that the bottom spacing is inconsistent between Hex and the two other controls (RGB and HSL)
4. Notice that RGB and HSL have a horizontal scrollbar if at least one range control slider is set to 100% (an easy way to test this is to set to white in RGB)
5. After this PR, there should be consistent bottom padding between color input types, and there should be no horizontal scrollbar

(Note: in Storybook, the `ColorPicker` component is not rendered in a Popover, so it's best to test this component change in the editor)

## Screenshots or screencast <!-- if applicable -->

| RGB Before | RGB After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/172970707-b6244632-82f6-41c7-ab89-1a1903734be9.png) | ![image](https://user-images.githubusercontent.com/14988353/172974043-0ef6e470-9359-4562-9be2-274c0c400f9c.png) |
| Hex Before | Hex After |
| ![image](https://user-images.githubusercontent.com/14988353/172974188-d54182df-ad42-4730-8a1d-c8da90c52040.png) | ![image](https://user-images.githubusercontent.com/14988353/172974071-435fc221-1d71-44a2-b26f-8e5ba553abf9.png) |

andrewserong committed:
Color Picker: Remove horizontal scrollbar when using HSL or RGB color input types

andrewserong committed:
Fix vertical spacing

andrewserong committed:
Add changelog entry

andrewserong commented:
Lena and Marco — I've added you both as reviewers, but there's no urgency with this one at all as I'll be finishing up for the week shortly and am AFK on Monday 🙂

glendaviesnz reviewed (approved):
Tested well for me.

ciampo commented:
Code changes LGTM 🎉  I'll ping @jameskoster in case he has the time to take a look from the design perspective too!

jameskoster commented:
It's mostly working well, but I do still see a horizontal scrollbar appearing intermittently (with scrollbars set to always display in macOS sys prefs):


https://user-images.githubusercontent.com/846565/173054421-d5b0d51d-90ae-4ccf-9bd0-509b70cbd107.mp4

Almost certainly unrelated, but another strange issue – when you drag the range to its minimum, the adjacent text input displays the maximum value.

mirka commented on packages/components/src/color-picker/hex-input.tsx:
```diff
@@ -49,23 +50,25 @@ export const HexInput = ( { color, onChange, enableAlpha }: HexInputProps ) => {
-			hideLabelFromVision
-			__unstableStateReducer={ stateReducer }
-		/>
+		<Spacer as={ HStack } spacing={ 4 }>
```
This hack feels weird to me, in that hex-input doesn't need to be `HStack`ed nor have `spacing` — this `Spacer` is only being used for the [default margin-bottom](https://github.com/WordPress/gutenberg/blob/2fa50a63742f7aace59d7fcb3d0aa41575bd7bbf/packages/components/src/spacer/hook.ts#L21) that comes with it.

This might have hidden complexity that I'm missing so it's not a blocker for this PR, but I feel like the CSS structure should be refactored to be more like this:

<img src="https://user-images.githubusercontent.com/555336/173138156-74e1614a-65b7-44bf-b634-308bb8659935.png" alt="Color Input using a padded and a flex gapped container" width="400">

This way we can use flex gap instead of the Spacer's margin-bottom, and don't have to wrestle with the final item's margin-bottom bleeding out and messing with our outermost padding.

mirka reviewed (commented):
FYI I believe `ColorPalette` is the component you're looking for in the Storybook for testing — it's the one that has `ColorPicker` wrapped in a popover ✨ 

> It's mostly working well, but I do still see a horizontal scrollbar appearing intermittently

Yes, it looks like the main color picker dot (`.react-colorful__pointer`) is overflowing the container. I looked into this and noticed this line here:

https://github.com/WordPress/gutenberg/blob/2fa50a63742f7aace59d7fcb3d0aa41575bd7bbf/packages/components/src/color-palette/style.scss#L29-L30

I _think_ this is what we want (vis-à-vis the dot, and maybe even the sliders?), and it's certainly the intent of the code. However, there's some JS somewhere that's adding an inline style that's overriding the `overflow`:

<img width="495" alt="div.components-popover__content with inline style overriding the overflow to auto" src="https://user-images.githubusercontent.com/555336/173144180-1d140922-028b-4290-8852-6b9eccaf2847.png">

So if you `overflow: visible !important`, it's fixed. I have a feeling this is a regression (cf. the screenshot in #36963). It would be nice to get to the root of this, but if all else fails, maybe we resort to `!important` 😅

andrewserong committed:
Switch to flex spacing, tidy up spacing

andrewserong commented:
Thanks for the reviews and feedback @mirka and @jameskoster, apologies for the delay — I switched over to some other tasks this week, but have just given this a rebase and tweaked some things a little further.

I've done the following:

* Removed the Spacer wrapper on the hex input
* Used Flex with gap to contain the sliders
* Added `overflow: hidden` to the `.react-colorful` selector to fix the other horizontal scrollbar issue — this seemed a bit easier than trying to wrestle with the Popover's settings. It looks like when the Popover was refactored to use floatingUI, the `overflow: auto` rule [was introduced here](https://github.com/WordPress/gutenberg/blob/3e9a9a891ceac9fcfaf4b3c3b61a1ee259f3809c/packages/components/src/popover/index.js#L193) — I think adding this rule might be okay as a pragmatic fix.
* Upped the bottom padding slightly to remove vertical scrollbar again — it turns out that the Tooltip used within the RangeControl component[ sets a `bottom: -80%` rule](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/components/src/range-control/styles/range-control-styles.js#L259) which forces the Tooltip on the bottom-most RangeControl to make the ColorPicker component overflow. I feel like this _shouldn't_ really happen — the position of the tooltip in the RangeControl component probably needs to be tweaked a bit to factor in if it's this close to the edge, but a quick fix of upping the bottom padding from `16px` to `20px` fixes it for now.

Here's how it's looking now:

<img width="243" alt="image" src="https://user-images.githubusercontent.com/14988353/174241201-45ec7e1d-3a1d-4345-94de-0c21cc915686.png">

Please let know if you think there's different / better trade-offs to be made, and I can follow up next week.

Thanks! 🙇

> Almost certainly unrelated, but another strange issue – when you drag the range to its minimum, the adjacent text input displays the maximum value.

Yes, that seems a bit weird, not too sure what's going on there. Happy to take a look at that separately.

jameskoster reviewed (approved):
This looks good to me and the scrollbar issue seems to have been fixed:


https://user-images.githubusercontent.com/846565/174292574-0d8f7be1-7a20-4c6d-b350-2b3c8dac6973.mp4

Another tangential observation, it's a bit strange how the palette popover has different styling to the custom color one. I guess I'll open an issue for that one :)

andrewserong commented:
Thanks for the follow-up review and for opening up the borders inconsistency in: https://github.com/WordPress/gutenberg/issues/41787 👍

I'll merge this one in now, and we can look at follow-ups separately 🙂
